linked_list/linked_list.py

```diff
@@ -9,71 +9,161 @@ def __init__(self, value, next_node = None):
-    # Space Complexity: ?
+    # Time Complexity: O(n) - traverses the whole list
+    # Space Complexity: O(1) - nothing new is being added, so memory is unaffected
     def search(self, value):
-        pass
+        current = self.head
+
+        while current:
+            if current.value == value:
+                return True
+            current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) - traverses the list
+    # Space Complexity: O(1) - no new node is being added, so memory is unaffected
     def length(self):
-        pass
+        current = self.head
+        # set variable to 0, to initiate counter
+        list_length = 0
+        while current != None: 
+            # add to counter 
+            list_length += 1
+            # set current to next to traverse
+            current = current.next
+        # return list_length when while loops breaks 
+        return list_length
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) - traverses the list to find value
+    # Space Complexity: O(1) - no new node is being added, so memory is unaffected
     def get_at_index(self, index):
-        pass
+        current = self.head
+        # return None if fewer nodes (can be equal!) than index value
+        if index >= self.length():
+            return None
+        count = 0
+        # check count versus index
+        while current != None: 
+            if count == index: 
+                return current.value
+            # set current to next 
+            current = current.next
+            # increment count
+            count += 1
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) - traverses list? 
+    # Space Complexity: O(1) - no new node is being added, so memory is unaffected
     def get_last(self):
-        pass
+        current = self.head
+        while current != None:
+            # if next is None, end of list, so return value
+            if current.next == None:
+                return current.value
+            else: 
+                # set current to next, if next != None
+                current = current.next
+        # if list is empty return None
+        return None
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1) - only inserts 1 value
+    # Space Complexity: O(n) - traverses the list in order to find the end, and then adds on value
     def add_last(self, value):

```

Comment:
👍  The time complexity is O(n) because you traverse to the end of the list.  The space complexity is O(1) because you only ever add one node.